app/src/main/java/org/p2p/wallet/home/ui/wallet/WalletPresenter.kt

```diff
@@ -24,58 +27,69 @@ class WalletPresenter(
     // mappers
     private val homeMapper: HomePresenterMapper,
     tokenKeyProvider: TokenKeyProvider,
+    private val tokenServiceCoordinator: TokenServiceCoordinator
 ) : BasePresenter<WalletContract.View>(), WalletContract.Presenter {
 
     private var username: Username? = null
 
-    private val buttonsStateFlow = MutableStateFlow<List<ActionButton>>(emptyList())
     private val refreshingFlow = MutableStateFlow(true)
 
     private val userPublicKey: String by unsafeLazy { tokenKeyProvider.publicKey }
 
     override fun attach(view: WalletContract.View) {
         super.attach(view)
         observeRefreshingStatus()
-        observeActionButtonState()
 
         loadInitialData()
+
+        observeUsdc()
     }
 
-    private fun observeActionButtonState() {
+    private fun observeUsdc() {
         launch {
-            buttonsStateFlow.collect { buttons ->
-                view?.showActionButtons(buttons)
+            tokenServiceCoordinator.observeUserTokens()
+                .collect { handleTokenState(it) }
+        }
+    }
+
+    private fun handleTokenState(newState: UserTokensState) {
+        view?.showRefreshing(isRefreshing = newState.isLoading())
+
+        when (newState) {
+            is UserTokensState.Idle -> Unit
+            is UserTokensState.Loading -> Unit
+            is UserTokensState.Refreshing -> Unit
+            is UserTokensState.Error -> view?.showErrorMessage(newState.cause)
+            is UserTokensState.Empty -> view?.showBalance(null)
+            is UserTokensState.Loaded -> {
+                val usdc = newState.solTokens.find { it.isUSDC }
```

Comment:
but we should get all tokens with needed extras not only USDC :(

Reply:
For crypto, right?
For Wallet you need only usdc